### PR TITLE
Iterate on user consent dialog for identity server

### PR DESCRIPTION
**Pull request by bmarty**
Implements #4577 

New dialogs: https://github.com/vector-im/element-android/issues/4577#issuecomment-982701319

We need to retrieved the T&C, that's why many changes are necessary. Sadly there is some code duplication between the ViewModels, we may find a way to reduce that later.

### CODE REVIEW

File: vector/src/main/java/im/vector/app/features/discovery/Extensions.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.
+ */
+
+package im.vector.app.features.discovery
+
+import im.vector.app.core.utils.ensureProtocol
+import org.matrix.android.sdk.api.session.Session
+import org.matrix.android.sdk.api.session.terms.TermsService
+
+suspend fun Session.fetchIdentityServerWithTerms(userLanguage: String): IdentityServerWithTerms? {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/discovery/Extensions.kt:23]: :100: for lifting out of the view model

File: vector/src/main/java/im/vector/app/features/contactsbook/ContactsBookViewEvents.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.contactsbook
+
+import im.vector.app.core.platform.VectorViewEvents
+import im.vector.app.features.discovery.IdentityServerWithTerms
+
+sealed class ContactsBookViewEvents : VectorViewEvents {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/features/contactsbook/ContactsBookViewEvents.kt:22]: worlds smallest nit: could be a sealed interface, saves having to write `()` in the extends :sweat_smile:

File: vector/src/main/java/im/vector/app/core/utils/Dialogs.kt [hunk 3]
```diff
@@ -40,16 +45,36 @@ fun Context.displayInWebView(url: String) {
             .show()
 }
 
-fun Context.showIdentityServerConsentDialog(configuredIdentityServer: String?, policyLinkCallback: () -> Unit, consentCallBack: (() -> Unit)) {
+fun Context.showIdentityServerConsentDialog(identityServerWithTerms: IdentityServerWithTerms?,
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/core/utils/Dialogs.kt:48]: general question about identity servers, are privacy policies enforced?

**Review by ouchadam — APPROVED** [review]
LGTM :+1:

File: vector/src/main/java/im/vector/app/features/contactsbook/ContactsBookViewEvents.kt (hunk 2 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/contactsbook/ContactsBookViewEvents.kt:22]: I think at the beginning of the project sealed interface was not a thing, but now, yes. I prefer to perform a global replace on a dedicated PR

File: vector/src/main/java/im/vector/app/core/utils/Dialogs.kt (hunk 3 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/core/utils/Dialogs.kt:48]: You mean can we have an IS without any privacy policy? I do not think so, but the code should be robust to that.
By the way I will create some issues (defect, but no P1) I have seen in the area

File: vector/src/main/java/im/vector/app/features/contactsbook/ContactsBookViewEvents.kt (hunk 2 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/contactsbook/ContactsBookViewEvents.kt:22]: (created https://github.com/vector-im/element-android/issues/4588 about it)

**Comment by bmarty** [comment]
Just pushed a small commit

**Pull request merged**